# ETL de australian_user_reviews

In [14]:
import ast
import pandas as pd
import herramientas as h

Importamos el archivo australian_user_reviews a un Dataframe

In [15]:
archivo = 'australian_user_reviews.json'

rows = []

with open (archivo, encoding='utf-8') as f:
    for line in f.readlines():
        rows.append(ast.literal_eval(line))

df_user_reviews = pd.DataFrame(rows)

In [16]:
df_user_reviews.head()

user_id                                           user_url   
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...  \
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                                             reviews  
0  [{'funny': '', 'posted': 'Posted November 5, 2...  
1  [{'funny': '', 'posted': 'Posted June 24, 2014...  
2  [{'funny': '', 'posted': 'Posted February 3.',...  
3  [{'funny': '', 'posted': 'Posted October 14, 2...  
4  [{'funny': '3 people found this review funny',...

In [17]:
df_user_reviews.tail()

user_id                                           user_url   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...  \
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...

Como podemos notar a primera vista, este Dataframe tiene columnas anidadas, por lo que procedemos a desanidarlas

In [18]:
unpacked_data = []
for index, row in df_user_reviews.iterrows():
    user_id = row['user_id']
    user_url = row['user_url']
    for review_dict in row['reviews']:
        review_dict['user_id'] = user_id
        review_dict['user_url'] = user_url
        unpacked_data.append(review_dict)

df_user_reviews_desanidado = pd.DataFrame(unpacked_data)

In [19]:
df_user_reviews_desanidado.head()

funny                     posted last_edited item_id   
0         Posted November 5, 2011.                1250  \
1            Posted July 15, 2011.               22200   
2           Posted April 21, 2011.               43110   
3            Posted June 24, 2014.              251610   
4        Posted September 8, 2013.              227300   

                                           helpful  recommend   
0                                   No ratings yet       True  \
1                                   No ratings yet       True   
2                                   No ratings yet       True   
3  15 of 20 people (75%) found this review helpful       True   
4     0 of 1 people (0%) found this review helpful       True   

                                              review            user_id   
0  Simple yet with great replayability. In my opi...  76561197970982479  \
1               It's unique and worth a playthrough.  76561197970982479   
2  Great atmosphere. The gunplay can be a bit chu...  76561197970982479   
3  I know what you think when you see this title ...            js41637   
4  For a simple (it's actually not all that simpl...            js41637   

                                            user_url  
0  http://steamcommunity.com/profiles/76561197970...  
1  http://steamcommunity.com/profiles/76561197970...  
2  http://steamcommunity.com/profiles/76561197970...  
3               http://steamcommunity.com/id/js41637  
4               http://steamcommunity.com/id/js41637

In [20]:
df_user_reviews = df_user_reviews_desanidado

Guardamos el Dataframe en formato CSV para poder trabajar mejor

In [21]:
df_user_reviews.to_csv('df_user_reviews.csv', index=False)

In [22]:
df_user_reviews = pd.read_csv('df_user_reviews.csv')

Ahora analizamos los valores nulos y duplicados

In [23]:
h.data_info(df_user_reviews)


Total rows:  59305

Total full null rows:  0

Total duplicated rows: 874


Column                             dType  No_Null_%  No_Null_Qty   
0        funny  [<class 'float'>, <class 'str'>]      13.74         8151  \
1       posted                   [<class 'str'>]     100.00        59305   
2  last_edited  [<class 'float'>, <class 'str'>]      10.35         6140   
3      item_id                   [<class 'int'>]     100.00        59305   
4      helpful                   [<class 'str'>]     100.00        59305   
5    recommend                  [<class 'bool'>]     100.00        59305   
6       review  [<class 'str'>, <class 'float'>]      99.95        59275   
7      user_id                   [<class 'str'>]     100.00        59305   
8     user_url                   [<class 'str'>]     100.00        59305   

   Null_%  Null_Qty  
0   86.26     51154  
1    0.00         0  
2   89.65     53165  
3    0.00         0  
4    0.00         0  
5    0.00         0  
6    0.05        30  
7    0.00         0  
8    0.00         0

Como hemos visto, el Dataframe no contiene valores nulos, pero tiene 874 filas duplicadas, esas filas sencillamente las eliminaremos.

In [24]:
df_user_reviews.drop_duplicates(inplace=True)

In [25]:
h.data_info(df_user_reviews)


Total rows:  58431

Total full null rows:  0

Total duplicated rows: 0


Column                             dType  No_Null_%  No_Null_Qty   
0        funny  [<class 'float'>, <class 'str'>]      13.71         8010  \
1       posted                   [<class 'str'>]     100.00        58431   
2  last_edited  [<class 'float'>, <class 'str'>]      10.33         6037   
3      item_id                   [<class 'int'>]     100.00        58431   
4      helpful                   [<class 'str'>]     100.00        58431   
5    recommend                  [<class 'bool'>]     100.00        58431   
6       review  [<class 'str'>, <class 'float'>]      99.95        58401   
7      user_id                   [<class 'str'>]     100.00        58431   
8     user_url                   [<class 'str'>]     100.00        58431   

   Null_%  Null_Qty  
0   86.29     50421  
1    0.00         0  
2   89.67     52394  
3    0.00         0  
4    0.00         0  
5    0.00         0  
6    0.05        30  
7    0.00         0  
8    0.00         0

Eliminamos los pocos nulos que quedaron en la columna 'review'

In [26]:
df_user_reviews.dropna(subset=['review'], inplace=True)

In [27]:
h.data_info(df_user_reviews)


Total rows:  58401

Total full null rows:  0

Total duplicated rows: 0


Column                             dType  No_Null_%  No_Null_Qty   
0        funny  [<class 'float'>, <class 'str'>]      13.71         8009  \
1       posted                   [<class 'str'>]     100.00        58401   
2  last_edited  [<class 'float'>, <class 'str'>]      10.34         6037   
3      item_id                   [<class 'int'>]     100.00        58401   
4      helpful                   [<class 'str'>]     100.00        58401   
5    recommend                  [<class 'bool'>]     100.00        58401   
6       review                   [<class 'str'>]     100.00        58401   
7      user_id                   [<class 'str'>]     100.00        58401   
8     user_url                   [<class 'str'>]     100.00        58401   

   Null_%  Null_Qty  
0   86.29     50392  
1    0.00         0  
2   89.66     52364  
3    0.00         0  
4    0.00         0  
5    0.00         0  
6    0.00         0  
7    0.00         0  
8    0.00         0

Terminamos de rellenar los nulos faltantes

In [28]:
h.reemplazar_nulos(df_user_reviews)

In [29]:
h.data_info(df_user_reviews)


Total rows:  58401

Total full null rows:  0

Total duplicated rows: 0


Column             dType  No_Null_%  No_Null_Qty  Null_%  Null_Qty
0        funny   [<class 'str'>]      100.0        58401     0.0         0
1       posted   [<class 'str'>]      100.0        58401     0.0         0
2  last_edited   [<class 'str'>]      100.0        58401     0.0         0
3      item_id   [<class 'int'>]      100.0        58401     0.0         0
4      helpful   [<class 'str'>]      100.0        58401     0.0         0
5    recommend  [<class 'bool'>]      100.0        58401     0.0         0
6       review   [<class 'str'>]      100.0        58401     0.0         0
7      user_id   [<class 'str'>]      100.0        58401     0.0         0
8     user_url   [<class 'str'>]      100.0        58401     0.0         0

Guardamos el Dataframe limpio en formato CSV

In [30]:
df_user_reviews.to_csv('australian_user_reviews.csv', index=False)